Remove charges from solid comestibles

mqrause opened the pull request:
#### Summary
Infrastructure "Remove charges from solid comestibles"

#### Purpose of change

Starting to phase out `count_by_charges`.
Fixes #47842: In general `count_by_charges` items should never have pockets, because a `hide bag (1)` has the same amount of pocket space as a `hide bag (2)` anyway. Might be worth it to add a check for that.

#### Describe the solution

- exclude solid items from `count_by_charges`
- items that aren't `count_by_charges` anymore get migrated, unless they get added to the charge removal list
- recipe `charges` work both for items with and without `count_by_charges`
- iuse_transform can transform into a container containing multiple target items, not just `count_by_charges` items
- Because solid comistibles can't have charges anymore, default charges of comestibles is now 0. This means liquids/gases need to explicitly define charges. As a side effect, solid comestibles can't copy-from from liquid/gaseous comestibles anymore.
- automated json adjustments with minor manual adjustments to fix test failures

Possible todos/issues:

1. include inline item groups in automated adjustments
   - it doesn't cause direct issues to leave them as is, but spawn amounts are wrong since charges get ignored
   - they're sprinkled everywhere in all kinds of other json objects
2. automatically created groups that end in `_inf`
   - The intention of those groups is to spawn an item in a container, but the container is filled with a random number between min and filling up the container. This behavior currently isn't replicable with `count` instead of `charges`, so it either needs code support or explicitly calculated max amounts for these groups.
3. ~~sealed container issues~~ should be mostly fixed by #66971 
   - ~~Spawning sealed containers seems to depend on the container being completely filled, with a remaining volume of 0 ml. The charge removal and volume changes break that for some items.~~
   - ~~for example 4 charges in 250ml put into a glass jar turns into 4 items of 62ml, so 248ml, which causes the glass jar to not be sealed anymore, because it's not full~~
4. tiny items <1ml
   - A few items had charges with a volume < 1ml. Since we don't support that they were rounded up to 1ml. I'll include a list of these items here once it's ready to merge.

Other observations:

- Things like `"charges": [ 0, 2 ]` can't actually spawn 0 charges from what I have seen. Since this got changed to `"count": [ 0, 2 ]`, this can actually spawn nothing now. Because some of these cases got moved to their own item groups due to container issues, a count of 0 in those cases means an empty container is spawned. This may or may not be what people intended with these groups, but there's not much to do about it with just automated changes.

#### Describe alternatives you've considered

Also including liquids/gases in this, but I don't wanna imagine the kinds of things that would break.

#### Testing

Tests pass, and old games seem to load fine, but only minor testing was done for now.

#### Additional context

Gist with some data from the automated json changes:
https://gist.github.com/mqrause/dc234737bfd5c4309aa4dc496d17bd53

anoobindisguise commented:
This is going to be horrible without some serious reworks to how non-charge items work. 
For example, take caffeine pills. The recipe makes nearly 2000, and when this happens, you will only be able to store 4096 of them (less than 1 liter) in any given place. It will take over 2 hours to fully empty a bottle with 4096 pills, or to insert 4096 pills into a bottle. Then there is the issue of 4096 pills being extremely laggy, when there are so many items nearby it takes 10 seconds or more to eat a single item. This is easily observable when you have a gigantic stockpile of food and it takes ages to eat anything from it, whereas if you have a stack of a few thousand slow spoilage food items in consolidated stacks - for example 800 necco wafers, 800 chocolate candies, 800 dehydrated meats, and 800 hard cheeses, it takes less than a 5th of a second to eat something. Turn those into individual items and the time to eat something will increase 1000 fold. There is also the extreme time it takes to move items around and the sheer clunkiness of navigating a food menu that is 100 pages long that could have been literally 1 page long. And the lag so many items calculating spoilage will generate. If we do this then every single food stockpile is going to be like the awful MRE warehouse in the mil_base. 
I don't think it's an exaggeration to say that this will make the game completely unplayable. It's probably less work to just fix the problems with charges than dealing with the madness that is uncharged comestibles. Individual tailoring patches are already bad enough, anyone who's tried to accumulate tailoring supplies currently can attest to that.

mqrause commented:
> the sheer clunkiness of navigating a food menu that is 100 pages long that could have been literally 1 page long.

Nothing will change in menus, except the number will be in the front instead of the back of the name.
The other issues need fixing anyway.

> It's probably less work to just fix the problems with charges than dealing with the madness that is uncharged comestibles.

There is no "just fix the problems with charges", they're horrible from an infastructure and maintainability standpoint with special cases and separate logic at every corner.

Zireael07 commented:
@anoobindisguise however raises some valid points. There's already at least one open issue about many items being laggy (thread, rags). This looks likely to exacerbate this...

mqrause commented:
It's not likely to, it absolutely will make these issues worse. They just need someone to work on them.

Amoebka commented:
Alright, so perhaps this PR should be on hold until those issues are fixed then? Otherwise you are just making life miserable for players, and reducing the number of people playing the latest experimentals, on whom the project depends for reporting bugs.
Similar situations have happened before, and making existing issues worse most definitely did NOT make the fixes arrive faster.

mqrause commented:
It's not for me to decide if and when this gets merged and it's kinda pointless to talk about it to this extent at this point in time.

chrispikula commented:
So what will this do to items that have more than 4096 charges, which is the item stack?  Just overflow in the transition?  If so, my vehicle storage is going to delete many, many items, as overflow is limited to, iirc, 3 or 4 tiles.
![image](https://user-images.githubusercontent.com/22281606/189499981-663ac45f-9b11-441b-af46-1eaf83ec2113.png)

And what about liquids in tanks?
![image](https://user-images.githubusercontent.com/22281606/189500040-110063a7-c531-4451-aa66-cb451504a070.png)
I don't see gasoline being made into a large unit volume any time soon.

mqrause commented:
As it _currently_ is, items are migrated to a single item with all the other charges as single items in its migration pocket, which isn't subject to the limit. This state will persist until you put the item in your inventory, where it will unload the migration pocket. I think currently you can even wield it without it unloading as long as you don't touch anything else in your inventory, so if you're smart about it, nothing will be lost. I still need to look at it later on because of the amount and nature of the items affected, since I'm not sure how it behaves when crafting/consuming. But that's just the finishing touches of the PR once everything else is in place.

chrispikula commented:
Or if you put something in your medication storage trunk, right after the update?  'cause that'll update that tile, yeah?

If it does, that will cause spillage, and thus item deletion when it can't spill anything else.  As it stands, my powdered food & adjacent goods storage has upwards of 82 thousand charges of items stored in it.  Beside it is the chemical storage has 248k charges of items in it.  Not including liquids.  That'll end up being upwards of 85 hours of picking up items, one by one, second by second, in game and trying to find new places to store them, if they don't flow into the 'can't place item' error.

Side note, possible issue is that gunpowder and all of the items that copy-from it come in a count of 4540 charges when you get them.  Will these just be expected to spill?  It's also gonna feel kind of like a waste when you can't place more than a single jar of gunpowder into a cargo hold.

mqrause commented:
_Currently_ **only** things in your inventory get updated. Map/vehicle tiles don't do anything.

As long as gunpowder comes in a container, there is no issue whatsoever. Containers are not affected by the item limit. And only the container itself counts towards the item limit. And frankly if this update stops people from storing their foodstuffs, chemicals and powders of any kind directly on the floor, it's mostly a good thing.

chrispikula commented:
Sure, yeah, if there was a way to do that without involving even more highly annoying, repetitive button pressing, I'd agree.  But we currently can't put things into pre-existing containers while they are in storage yet, other than fluids.  

How will this deal with merging multiple partially filled containers?  Will it be rate-limited to the roughly 1 second per item transfer rate of normal items?  Given a lack of bulk movement transfer, that is.  Common cases of this would be all of the 100-units of salt you find in kitchens, that you'd probably want to put into a single container so as to not have half your storage wasted by small plastic bottles and vials.

Uncommon case would come from disassembling a car battery for lead, where you'd get 4 thousand or so.  Moving that off the workbench beside you would take some time, yeah?

mqrause commented:
As said before, that's explicitly not a concern of this PR. There are issues for that kind of stuff already (and if not feel free to open one). If the main dev team deems those issues critical for this PR, it won't be merged until they're closed and that's that.

mqrause commented:
Rebased on #62498 because that's required for this work properly. Not quite ready for playtesting and I still wanna rework recipes a little to allow for more flexibility, but it mostly works if you ignore all the items being broken.

Commit:
adjust iuse_transform

mqrause commented:
This is "done", I think. Pending all the json changes to items and recipes and documentation updates.
Slightly reorganized the commits to make it possibly easier to review, but without testing and without the json changes they don't really work by themselves.

mqrause commented:
Thanks @Reaper-Of-Souls for the json changes!

Ilysen commented:
I'm a bit concerned about the issues that other people have raised in this PR that seem like they're not being addressed. anoob raises very valid points that some of the very high-amount crafts and items in the game will take an extremely long time to move around if this PR gets merged, and so far I feel like that hasn't received an answer besides "someone just needs to work on the problem". Exacerbating that existing issue seems like a step backwards to me; why make it worse when you haven't yet fixed the underlying issues? Isn't that rather like taking off a bandage from a bleeding wound and then stabbing yourself?

This PR would raise issues with item limits and movement time that seem like are not being addressed. "It seems like a good thing if it prevents people from storing items outside containers" is not, in my eyes, a valid excuse, because it provides a laggy and unclear incentive that the majority of people will only be thrown off by because it's caused by an engine limitation and not by any in-game means.

I guess to put a bow on all this: what will this PR offer to your average player that makes the under-the-hood changes worthwhile? What is the benefit of doing this now as opposed to waiting until existing issues are fixed to avoid exacerbating existing issues? I think `"It's not for me to decide when this gets merged"` is a fallacy; if you're making the PR, and advocating for its changes to get in, then you have to take responsibility for what happens as a result of it. If the changes would not happen if you did not make them, then I think you need to be able to provide answers to these questions; you can't defer to "the dev team" for an answer when you're the one writing the code.

mqrause commented:
This PR should be in a state that allows actually testing all your concerns and how big the impact will actually be. If you think the impact is big enough to warrant postponing it, then I recommend you collect some data on it.

However since those aren't new issues at all, and nobody tried to do anything about it yet, waiting for them to get fixed is basically the same as abandoning this PR.

Ilysen commented:
I don't feel like that answers any of my questions. It's just deferring responsibility onto other folks again. If the answer is "the data isn't there", then what is this PR meant to address? I *could* go in-game and collect the data showing that a ton of charges of meat jerky and a ton of individual units of meat jerky are miles apart in terms of slowdown and clunkiness, but would that actually change your mind? I feel like it probably wouldn't, but I can get it if you insist. If you've played with the new tailoring system recently, I imagine you'll be well aware of how slow it is to move hundreds and hundreds of individual patches and sheets from place to place.

> However since those aren't new issues at all, and nobody tried to do anything about it yet, waiting for them to get fixed is basically the same as abandoning this PR.

I feel like given all the existing concerns and problems that this PR amplifies, abandoning this would in fact be preferable to it being merged. If the underlying issues are such a problem, why not address those, and not the symptom that results from them?

Zireael07 commented:
I agree with @Ilysen, I recall at least two issues showing that tons of individual items are very slow to operate on unlike charges

NetSysFire commented:
E.g moving 1337 handful of leaves or cotton patches.

mqrause commented:
I'm fully aware of the issues. This PR isn't meant to address them and it also can't address them. It addresses a plethora of other issues (not just for players) that are introduced by charges, which I'm sure could use some better communication.

It's completely fair to be bothered by those performance issues and see them as worse than the issues addressed by this PR. And it's also completely fair to appeal to the dev team to wait for those issues to be fixed before merging this. But I find it quite offensive you expect me to fix those instead of working on this, Ilysen.

Ilysen commented:
I don't have any desire to cause personal friction, and so I'm sorry that I have; I'm trying to be direct, not cold. I think that directness is important, though, in this particular case. this is a change that will affect everyone that plays the game by causing significant slowdown and further issues for no front-end improvements. what else am I supposed to be advocating for here? that it's fine to introduce heavy performance regressions for everyone that plays the game because the underlying problem hasn't been fixed yet? I feel like the underlying cause \*should\* be solved first, not the imperfect fix (item charges) that mitigates the symptoms, and I don't feel like it's unreasonable to try and push for that to happen

mqrause commented:
The issue is not your performance concerns but all the other points you derive from that. This is an infrastructure PR, and as such all player facing changes are secondary (even if prominent in this case). The goal is to make the code easier to work with, enabling further development and preventing bugs before they appear. So no matter how valid the concerns for the performance impact are, this is the wrong place to push for fixes for that, because it's a completely separate issue.

Ilysen commented:
I'm aware of the positive health impacts that this PR would have on the game, but I'm just not convinced that they would outweigh the heavy negatives that would come as a result of it. if "PR that would greatly amplify existing flow issues with the game" is not a proper place to raise concerns about performance impacts, what \*is\* a proper place? an infrastructure change it may be, but it has heavy and prominent front-end impacts for everyone who plays the game; is that supposed to just *not matter* because it's made in service towards a longer-term project?

I find it impossible in my head to cleanly separate this PR from the "secondary" impacts it will have on front-end play, because as-is it can and will cause heavy usability issues for everyone, and I worry that that will make it much harder to actually play the game. you say yourself that those underlying issues haven't been fixed (and are in fact entirely valid!) and I don't know that this PR would do anything to speed that along. wouldn't even just a little disclaimer in the `Additional context` section mentioning that this is a long-term thing be enough?

NoddingCrow commented:
Just gonna swing by to say I agree this PR should not be merged until the core stacking issues are addressed by someone. The slowdown with tailoring individual items is already immense, and doing it with food too scares the pants off me. The idea of moving my salt (8000) as individual items on pickup, with individual pocket access times, is horrifying. 

That said, as an infrastructure change I support this draft. Dev team have been grumping about charges since time immemorial, so hey good on you for addressing. I agree with @Ilysen that a bit of context acknowledging the stacking issue and saying "This should not be merged until stacking is fixed" or words to that effect would go a long way to assuring people their games aren't about to be sent to computer hell, though -- which as far as I can tell is the chief source of opposition to this from the previous commenters.

mqrause commented:
> what *is* a proper place

The already existing issues for that like #58793.
It is not a new issue, it already affects the majority of items in the game. This simply expands that list of affected items to foodstuffs.

> wouldn't even just a little disclaimer in the `Additional context` section mentioning that this should be a long-term thing be enough?

From my perspective it isn't a long-term thing. I fully expect this to get merged within the first weeks of the new experimental. I also have heard no opinions from the dev team indicating otherwise. To quote erk for example: "The best time to break everything is right after stable releases anyway"

Zireael07 commented:
Even if this PR improves some aspects of the game, I think it shouldn't be merged because it does the improvement at a cost of exacerbating already awful performance issues (I tend to play Cata on an ancient computer, from a pendrive)

Qrox commented:
Even if we phased out `count_by_charges`, eventually we'll want to reduce memory consumption of large items stacks by merging identical items and storing the number of these items instead, which is more or less similar to what `count_by_charges` is currently doing, making most of these changes futile, so I think this is not the correct way ahead, not even as an intermediate state.

If we want a consistent system AND performance, my suggestion is to move forward to using charges everywhere for identical items, even though that would also require huge overhaul of the codebase.

mqrause commented:
> If we want a consistent system AND performance, my suggestion is to move forward to using charges everywhere for identical items, even though that would also require huge overhaul of the codebase.

I think charges in that sense shouldn't be a property of the item but be handled by whatever structure stores the items. That way all the relevant code changes live in the few places where items are stored/accessed. If we were to use the current system of charges, we would have to more or less rewrite all code that handles items in some way. Though it's hard to really estimate the impact that would have, I really can't imagine that being a more realistic approach.

Kamayana commented:
One thought I've had is that a lot of the charge items really should be treated similar to liquids. It's already bizarre that the player is encouraged to dump salt and detergent directly into their bags to be more volume efficient; it'd be even weirder for the player character to automatically distribute single grains of salt between all their pockets. Treating them like a liquid where they'd need their own container (including boxes, unlike liquid) would make a lot more sense.

mqrause commented:
Yes, handling powders more realistically is also something that's wanted but nobody picked up yet. Not directly related to this PR, though.

esotericist commented:
so i'm a little bit late and i apologize for interjecting after a month, but there's a pattern here that tells me that some clarity is needed:

for an infrastructure PR that is cleaning up the code, exacerbating existing performance issues is not only acceptable, it's _expected_. it's frequently difficult to fix those problems until after you have things working in a consistent, sane fashion.

also remember that if this goes in (and it might not, which depends on code review from folks who've had other priorities recently, esp going into a stable), it'd probably be early in an experimental cycle, and if you're playing experimental, you've **_opted into_** inconveniences like "terrible performance for in-progress development". further, frequently performance problems don't get fixed until they've become pressing enough that someone sits down and makes it their priority. making it more prevalent tends to help with that.

> an infrastructure change it may be, but it has heavy and prominent front-end impacts for everyone who plays the game; is that supposed to just not matter because it's made in service towards a longer-term project? you call it a separate issue, but is it really?
> 
> I find it impossible in my head to cleanly separate this PR from the "secondary" impacts it will have on front-end play, because as-is it can and will cause heavy usability issues for everyone, and I worry that that will make it much harder to actually play the game. 

that separation is crucial, though; it has become the policy of the project that things like that are not reasons not to fix things, they're reasons to fix things and then fix more things. if we wait to fix things for other fixes nobody has done yet, fixes don't happen. the project tried that approach in the past, and it resulted in things stalling that really needed doing.

in more recent releases, the pattern has been that we just push forward, fix the infrastructure, then when it's clear the attendant secondary issues are severe enough, they become blockers before stable, and that's likely to remain the approach. this is especially the case for infrastructure changes which are merely providing additional triggers for known performance problems, rather than more crucial problems (such as data integrity).

i hope everyone is able to keep this in mind for this and other similar PRs in the future, so as to limit derailing in focused PRs like this one.

kevingranade commented:
Echoing esotericist here, this is the correct scope for this PR.
It's not just that mqrause is allowed to keep the scope where it is, a larger scope would be wrong because it would make the PR too big.
If you want to pull this PR locally and test for severe breakage, please do so and let us know what happens, but it very much looks like this will be good to go immediately after the stable release.
If you want to mitigate issues uncovered by or exacerbated by this change, you can, but that is not a blocker for merging this PR post-release unless something makes the game literally unplayable.

mqrause commented on data/json/recipes/other/medical.json:
```diff
@@ -303,7 +306,8 @@
     "byproduct_group": [ { "item": "seed_cotton_boll", "charges": 2 } ],
-    "components": [ [ [ "cotton_boll", 1 ] ] ]
+    "components": [ [ [ "cotton_boll", 1 ] ] ],
+    "count":  10
```
Gonna ignore this for now cause json changes will be regenerated once other PRs get merged anyway.

andrei8l commented:
One major problem I see is that container names are no longer informative, especially now that they are collapsed by default (#62444)

![Screenshot from 2023-02-13 12-35-54](https://user-images.githubusercontent.com/68240139/218435643-3ae1740e-5009-4fb5-b5b9-6529105d8623.png)

Any ideas for fixing that? Parsing through all the contents is most likely prohibitive

esotericist commented:
> One major problem I see is that container names are no longer informative, especially now that they are collapsed by default (#62444)
> 
> ![Screenshot from 2023-02-13 12-35-54](https://user-images.githubusercontent.com/68240139/218435643-3ae1740e-5009-4fb5-b5b9-6529105d8623.png)
> 
> Any ideas for fixing that? Parsing through all the contents is most likely prohibitive

short term? no good answer, i expect. long term? i believe this falls into the same bucket as the earlier discussions, where the things that should be treated as a collective entity (fluid-adjacent entities, powders, or other things that you would treat as a single unit with volume or some other metric besides pure count) should be tracked by the containing structure.

notably, i don't think it's coherent behavior to collapse such container names by default if they have more than one kind of item, so we'll want to be doing something to check for type counts at the container level anyway.

andrei8l commented:
[Bethany Beach.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/10723445/Bethany.Beach.zip) This save is impossible to load with this patch. Started from the overrun scenario and stacked all the items in a corner so I can test an inv_ui perf patch and now I can't reload it without running out of memory (32Gb...)

mqrause commented:
Thanks, I'll look into it, though I'm not sure what would have changed so dramatically.

mqrause commented:
Okay, seems like the script for updating the json didn't account for `abstract` comestibles, so things copying `"abstract": "spice"` get migrated to items with charges, which get migrated again on next load to items with charges again etc. So it's a simple json issue.

Commit:
enforce proper charge definitions

Commit:
charge removal:
- remove charges from solid comestibles
- default to migrating charges if an item shouldn't have any
- allow consuming comestibles without charges

Commit:
remove copy-from from spread_peanutbutter

Commit:
keep test items stackable

Commit:
allow recipe charges for non-charge items

Commit:
misc fixes

Commit:
charge removal script

Commit:
fix pickle spawn amount

RenechCDDA commented:
I have reviewed the json changes at the current commit (https://github.com/CleverRaven/Cataclysm-DDA/pull/60885/commits/cc27609d8f0db38ab2b4cf4cddfe07425354215b), conversation took place on the development discord as @mqrause was online and responsive. I will not transcribe it here because it is far, far too long and I'm not bothering to format it.

My overall impression was that the changes look good - we identified some possible errors in the automated changes (very few for the number of ways items can be defined!) and edge cases and they're working towards resolving them. 

Conversation starts here on the dev discord: https://discord.com/channels/598523535169945603/598535827169083403/1102397996144001064

Commit:
fungal_seeds AMMO -> COMESTIBLE

Commit:
overflow items before consuming/using them to avoid removing items with migrated charges in their migration pocket

SurFlurer commented:
I'm trying to test this, and I noticed that the migration for comestibles stored in vehicle cargo space is struggling. They just lose all their charges and are turned into a single item. Example screenshot ( normally impossible to have a none-full sealed 3L jar ): 

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/22912139/c45f2ae6-766e-44cf-a75a-7f1682ed9423)

mqrause commented:
I think spilling items not on character would have a pretty bad performance impact, because it'd have to regularly check all the items. Best I can think of would be somehow displaying items inside migration pockets, but since I already fixed consuming such items (at least hope I caught all the cases of that), I don't really find it worth bothering with it as part of this because it's only a temporary thing anyway.

It could be done as a separate PR however, since it'd presumably also help with the "help, my collapsible item is gone after I un-/collapsed it while it was in a pocket that can only hold the item in the other state" issues.

SurFlurer commented:
You have the point. I'm compiling to test if it's possible to spill out from migration pockets at `map::actualize()`. I hope this will not cause so much burden. Also, is it possible to adapt item migration code to handle this overflow?

SurFlurer commented:
I've succeeded in overflowing all items in vehicle cargo space. But it takes several minutes to fully overflow hundreds of thousands items.

andrei8l commented:
> In-game performance is not visibly slower than without this change:
> 

Ok, now try with a 10x20 doomtruck of cargo spaces filled with items

SurFlurer commented:
> > In-game performance is not visibly slower than without this change:
> 
> Ok, now try with a 10x20 doomtruck of cargo spaces filled with items

Yes, I'm playing with [Emerald.Lake.Hills-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/11718337/Emerald.Lake.Hills-trimmed.tar.gz) ( the save I used to catch performance hogs ). It's not so big as 10x20, but is already 24 ton in weight.

mqrause commented:
One issue I see with this approach is the item limit per tile. Since it overflows everything, if you have multiple extremely large stacks of items, quite a bunch of the items might be deleted because there's no more room for them to be.

SurFlurer commented:
Yeah I've seen that happening. Tens of thousands of items fell onto the ground. But they'll flow into adjacent tiles instead of simply disappearing, right?

mqrause commented:
Max overflow range is 2 tiles, so there is a limit.

mqrause commented:
Not exactly the best time to rebase with #66273 being an issue, so I didn't test it beyond compiling and new comestibles since the last rebase might have some minor issues. Hope it helps people testing and trying it out anyway.

SurFlurer commented:
Anyway I made an all-overflowed version of Emerald Lake Hills, which can imitate the eventual state of migrated saves. If anyone would like to see how much performance is lost, they can compile then load this save:


[Emerald Lake Hills-overflowed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/11781903/Emerald.Lake.Hills-overflowed.tar.gz)

mqrause commented:
For the people notified for reviews, the push with the conflict resolutions is currently underway but pretty slow.

Maleclypse commented:
So I've loaded this up and I'm not hitting any errors.  I've noticed a number of places where changes have silently occurred without a fail but the wrong number of items are spawning.   That may be intended for now with a follow-up python script?
```
{ "item": "tuna_noodle_casserole", "charges": 8, "container-item": "casserole", "prob": 10 },
  628            { "item": "hot_corn_casserole", "charges": 8, "container-item": "casserole", "prob": 10 },
  629:           { "item": "cheesy_spinach_casserole", "charges": 8, "container-item": "casserole", "prob": 5 },
  630            { "item": "lazy_lobster_casserole", "charges": 8, "container-item": "casserole", "prob": 2 },
  631            { "item": "pot_copper", "prob": 5 },
```
All examples of these items I found were spawning a single item which again may be intended.  I'm assuming there are a ton of other ones that I suspected but couldn't identify off the top of my head.

mqrause commented:
> So I've loaded this up and I'm not hitting any errors. I've noticed a number of places where changes have silently occurred without a fail but the wrong number of items are spawning. That may be intended for now with a follow-up python script?
> 
> ```
> { "item": "tuna_noodle_casserole", "charges": 8, "container-item": "casserole", "prob": 10 },
>   628            { "item": "hot_corn_casserole", "charges": 8, "container-item": "casserole", "prob": 10 },
>   629:           { "item": "cheesy_spinach_casserole", "charges": 8, "container-item": "casserole", "prob": 5 },
>   630            { "item": "lazy_lobster_casserole", "charges": 8, "container-item": "casserole", "prob": 2 },
>   631            { "item": "pot_copper", "prob": 5 },
> ```
> 
> All examples of these items I found were spawning a single item which again may be intended. I'm assuming there are a ton of other ones that I suspected but couldn't identify off the top of my head.

The issue with that is that it's inline groups within groups and my script isn't capable of dealing with that. So it's intentional in the sense that I have to draw the line somewhere or I'll never finish this. The line isn't static, though, so if necessary I can put in more work to also fix cases like these.

Maleclypse commented:
> > So I've loaded this up and I'm not hitting any errors. I've noticed a number of places where changes have silently occurred without a fail but the wrong number of items are spawning. That may be intended for now with a follow-up python script?
> > ```
> > { "item": "tuna_noodle_casserole", "charges": 8, "container-item": "casserole", "prob": 10 },
> >   628            { "item": "hot_corn_casserole", "charges": 8, "container-item": "casserole", "prob": 10 },
> >   629:           { "item": "cheesy_spinach_casserole", "charges": 8, "container-item": "casserole", "prob": 5 },
> >   630            { "item": "lazy_lobster_casserole", "charges": 8, "container-item": "casserole", "prob": 2 },
> >   631            { "item": "pot_copper", "prob": 5 },
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > All examples of these items I found were spawning a single item which again may be intended. I'm assuming there are a ton of other ones that I suspected but couldn't identify off the top of my head.
> 
> The issue with that is that it's inline groups within groups and my script isn't capable of dealing with that. So it's intentional in the sense that I have to draw the line somewhere or I'll never finish this. The line isn't static, though, so if necessary I can put in more work to also fix cases like these.

It won't be necessary I believe.  I do hope that you make an issue on how to find these ones that can't be dealt with easily in the scope of this PR so that we can track and work on finishing off the leftovers, once this is merged.

mqrause commented:
Yes, I was planning to add some followup issues. Although "finding the ones that can't be dealt with easily" is the whole issue with them. Inline groups are scattered everywhere, from professions to item groups themselves to map spawns and probably other places. And the same goes for old style groups and spawns.

Commit:
automated json updates - items, itemgroups, and recipes

Commit:
post script fixes

mqrause commented:
Oops, sorry about that, just a slight rebasing error.

Maleclypse commented:
I'm guessing this is probably something going on with my local copy but this is the only errors I found and it was on start. 
```
23:28:18.956 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 1
23:28:23.785 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 2
23:28:23.788 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 3
23:28:23.788 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 4
23:28:23.789 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 5
23:28:23.789 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 6
23:28:23.791 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 7
23:28:23.791 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 8
23:28:23.792 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 9
23:28:23.792 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 10
23:28:23.794 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 11
23:28:23.794 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 12
23:28:23.796 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 13
23:28:23.796 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 14
23:28:23.798 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 15
23:28:23.798 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 16
23:28:23.800 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 17
23:28:23.800 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 18
23:28:23.802 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 19
23:28:23.802 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 20
23:28:23.805 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 21
23:28:23.805 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 22
23:28:23.807 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 23
23:28:23.807 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 24
23:28:23.810 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 25
23:28:23.810 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 26
23:28:23.823 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 27
23:28:23.823 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 28
23:28:23.825 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 29
23:28:23.825 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 30
23:28:23.826 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 31
23:28:23.826 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\mission.cpp:300 [assign] strange: player is already assigned to mission 32
```
If you aren't experiencing this error on start let me know and I'll go ahead and merge.  Ping me in discord to let me know.

alp7292 commented:
good now we have to wait 1 month (or more) untill all the performance issues and bugs fixed

Zireael07 commented:
It's sad that my concerns about performance were brushed aside only to be validated after this got merged :(

SurFlurer commented:
Eventually when pocket contents' total weight and volume can be cached, there shouldn't be too much difference.
